sdks/python/apache_beam/pipeline_test.py

```diff
@@ -528,6 +528,58 @@ def f_c(x):
       p.run().wait_until_finish()
     assert count_b == count_c == 4
 
+  def test_no_partial_writeouts(self):
+    from collections import defaultdict
+    from apache_beam.runners.direct.transform_evaluator import _TransformEvaluator
+    from apache_beam.runners.direct.transform_evaluator import _GroupByKeyOnlyEvaluator
+    from apache_beam.runners.direct.evaluation_context import _ExecutionContext
+
+    class TestTransformEvaluator(_TransformEvaluator):
+
+      def __init__(self):
+        self._execution_context = _ExecutionContext(None, {})
+
+      def start_bundle(self):
+        self.step_context = self._execution_context.get_step_context()
+        self.step_context.clear_partial_states()
```

Comment:
<!--new_thread; commit:af94b7b7a8c0ac0ada222884101ec10127f5b234; resolved:0-->
I think it would be better to have a `reset()` method on `_ExecutionContext` that can be called to clear the step context and remove `clear_partial_states()`.  We can then call this in `attempt_call()`.